sklearn/model_selection/_split.py

```diff
@@ -1101,6 +1117,28 @@ def split(self, X, y=None, groups=None):
                     indices[test_start : test_start + test_size],
                 )
 
+    def get_n_splits(self, X=None, y=None, groups=None):
+        """Returns the number of splitting iterations in the cross-validator
+        Parameters
+        ----------
+        X : object
+            Always ignored, exists for compatibility.
+
+        y : object
+            Always ignored, exists for compatibility.
+
+        groups : object
+            Always ignored, exists for compatibility.
+
+        Returns
+        -------
+        n_splits : int
+            Returns the number of splitting iterations in the cross-validator.
+        """
+        if self.n_splits == "walk_forward":
+            return X.shape[0] - (self.max_train_size + self.test_size) + 1
```

Comment:
Am I missing something here?

If `n_splits="walk_forward"`, `X.shape[0]=20`, `self.max_train_size=10`, and `self.test_size=5` we get a calculated `n_splits` of `20 - (10 + 5) + 1 = 6`. Thus across all splits we have a total test_size of `4 * 6 = 24` which is greater than our number of samples.

Is it not the expectation that these inputs would yield indices similar to the following?
```
[0 1 2 3 4 5 6 7 8 9] [10 11 12 13 14]
[5 6 7 8 9 10 11 12 13 14] [15 16 17 18 19]
```

Reply:
Yep we should return:

```python
return (X.shape[0] - self.max_train_size) // self.test_size
```

Reply:
The new equation only works for the case when the minimum train size is the maximum train size. In the docstring for `tscv = TimeSeriesSplit(n_splits='walk_forward', max_train_size=10, test_size=3)` it is shown that this is not the case. 

For the case in the docstring we calculate `(15 - 10) // 3 = 1`, though the examples given show that we would expect 4. See transcription below (this is why tests are failing currently):

```
    Fold 0:
      Train: index=[0 1 2]
      Test:  index=[3 4 5]
    Fold 1:
      Train: index=[0 1 2 3 4 5]
      Test:  index=[6 7 8]
    Fold 2:
      Train: index=[0 1 2 3 4 5 6 7 8]
      Test:  index=[ 9 10 11]
    Fold 3:
      Train: index=[ 2  3  4  5  6  7  8  9 10 11]
      Test:  index=[12 13 14]
```

The calculation should instead be `(X.shape[0] - self.min_train_size) // self.test_size`, where perhaps the default for `self.min_train_size` should be `self.test_size` for docstring accuracy. We could also allow for setting `self.min_train_size` on initialization (to allow for similar behavior to [#24589](https://github.com/scikit-learn/scikit-learn/pull/24589)), which would allow users to force each training batch to be of equal size for consistent CV scores on parameter hypotheses.